src/Processors/Formats/RowInputFormatWithDiagnosticInfo.cpp

```diff
@@ -65,30 +69,49 @@ String RowInputFormatWithDiagnosticInfo::getDiagnosticInfo()
 
     /// Roll back the cursor to the beginning of the previous or current row and parse all over again. But now we derive detailed information.
 
-    if (offset_of_prev_row <= in->buffer().size())
+    if (!is_errors_record && offset_of_prev_row <= in->buffer().size())
     {
         in->position() = in->buffer().begin() + offset_of_prev_row;
 
-        out << "\nRow " << (row_num - 1) << ":\n";
-        if (!parseRowAndPrintDiagnosticInfo(columns, out))
-            return out.str();
+        out_diag << "\nRow " << (row_num - 1) << ":\n";
+        if (!parseRowAndPrintDiagnosticInfo(columns, out_diag))
+            return std::make_tuple(out_diag.str(), out_data.str());
     }
     else
     {
         if (in->buffer().size() < offset_of_current_row)
         {
-            out << "Could not print diagnostic info because parsing of data hasn't started.\n";
-            return out.str();
+            out_diag << "Could not print diagnostic info because parsing of data hasn't started.\n";
+            out_data << "Could not collect raw data because parsing of data hasn't started.\n";
+            return std::make_tuple(out_diag.str(), out_data.str());
         }
 
         in->position() = in->buffer().begin() + offset_of_current_row;
     }
 
-    out << "\nRow " << row_num << ":\n";
-    parseRowAndPrintDiagnosticInfo(columns, out);
-    out << "\n";
+    char * data = in->position();
+    while (*data != '\n' && *data != '\r' && *data != '\0' && data < in->buffer().end())
```

Comment:
We should move `data < in->buffer().end()` to the first place in condition to prevent out-of-range.